data/README.md

```diff
@@ -1,132 +1,199 @@
 
-### `wp.data.registerActions( reducerKey: string, newActions: object )`
+select( 'my-shop' ).getPrice( 'hammer' );
+```
 
-If your module or plugin needs to expose its actions to other modules and plugins, you'll have to register action creators.
+#### `dispatch( storeName: string ): Object`
 
-An action creator is a function that takes arguments and returns an action object dispatch to the registered reducer to update the state.
+Given the name of a registered store, returns an object of the store's action creators. Calling an action creator will cause it to be dispatched, updating the state value accordingly.
 
-#### Example:
+_Example:_
 
 ```js
-wp.data.registerActions( 'myPlugin', {
-	setTitle( newTitle ) {
-		return {
-			type: 'SET_TITLE',
-			title: newTitle,
-		};
-	},
-} );
+const { dispatch } = wp.data;
+
+dispatch( 'my-shop' ).setPrice( 'hammer', 9.75 );
 ```
 
-### `wp.data.select( key: string )`
+#### `subscribe(): Function`
 
-This function allows calling any registered selector. Given a module's key, this function returns an object of all selector functions registered for the module.
+Given a listener function, the function will be called any time the state value of one of the registered stores has changed. This function returns a `unsubscribe` function used to stop the subscription.
 
-#### Example:
+_Example:_
 
 ```js
-wp.data.select( 'myPlugin' ).getTitle(); // Returns "My post title"
+const { subscribe } = wp.data;
+
+const unsubscribe = subscribe( () => {
+	// You could use this opportunity to test whether the derived result of a
+	// selector has subsequently changed as the result of a state update.
+} );
+
+// Later, if necessary...
+unsubscribe();
 ```
 
-### `wp.data.dispatch( key: string )`
+### Higher-Order Components
 
-This function allows calling any registered action. Given a module's key, this function returns an object of all action creators functions registered for the module.
+A higher-order component is a function which accepts a [component](https://github.com/WordPress/gutenberg/tree/master/element) and returns a new, enhanced component. A stateful user interface should respond to changes in the underlying state and updates its displayed element accordingly. WordPress uses higher-order components both as a means to separate the purely visual aspects of an interface from its data backing, and to ensure that the data is kept in-sync with the stores.
 
-#### Example:
+In the React/Redux ecosystem, this distinction is often labelled [_Presentational and Container Components_](https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0). [React Redux](https://github.com/reactjs/react-redux) is the official bindings between React and Redux for providing this behavior. The higher-order components offered in the data module are very much like React Redux's `connect` function, but offered as independent functions for retrieving and manipulating data: `withSelect` and `withDispatch`.
```

Comment:
I personally don't think it's necessary to explain Presentational and Container components, I feel it introduces more confusion than it solves.

Reply:
>I personally don't think it's necessary to explain Presentational and Container components, I feel it introduces more confusion than it solves.

What confusions do you have concerns about? I think the article goes into much more detail about this distinction and could be good for further reading, and to preempt thoughts of "Hey, that sounds a lot like...".

Reply:
I know for a lot of people, this distinction is important but from mine, it doesn't make a lot of sense :). 

A person reading this could assume that a Presentation component can be reused with any data or store, just providing props but that's not true because Presentation components can use Container components which means they are not reusable.

I like to think of everything as just React Components :). If I were to split types for React Components, I would prefer "Context-free components" and "Context-dependent components" :)

Reply:
For the purposes of this document in introducing someone who may have no familiarity of these concepts, how can we set them on the correct course? Is a "presentational and container" distinction harmful? Or can it be seen as more of a guideline to encourage developers from tying their components too directly to a specific source of props?

Reply:
So for me I think we can talk about `connect` as a Higher Order Component to inject props without the need to talk about Presentational and container components. Just explaining that a Higher Order Component is a function injecting extra props to a component and returning component.

I don't think you need to know Presentational and Container components to understand how HoC work.

Reply:
Okay, I think we could just remove it in the initial push.

One common pitfall I see where this distinction can be helpful is when component authors provide complex data unnecessarily to a component. Consider a `<Title />` component where someone passes the `post` object and we render via the value of `post.title`. Why does that component need the entire `post` object? It doesn't; it's less reusable, and on a more practical level it causes issues with unnecessary re-renders via pure-render on the object (the value of the post object is likely to change more frequently than the `title` string). The component author should just pass in the `title` string.

Thinking that guides like this can be useful for helping avoid these pitfalls, even if the rules themselves shouldn't be so strictly binding.

Reply:
Agree with the pitfall here, but not sure it's related to presentation/container separation, I know it's often explained when talking about this separation though.

I approved the PR, you choose :). I do have concerns about this separation but don't have strong objections to include it in the docs as it's a common React pattern people should know about.

Reply:
I'm going to drop this paragraph. The previous paragraph has a brief mention of separating visual and data concerns, which I think is sufficient for now.